### DISCUSSION TITLE
How I do pagination with this MS Graph API?

### DISCUSSION THREAD

**Discussion by jkroepke** [grafana/grafana-infinity-datasource#783] [Q&A]
Hi,

I'm trying to use MS Graph API with infinity datasource. I have currently issues to implement pagination.

```json
{
  "@odata.count": 500,
  "@odata.nextLink": "<full-url-with-specific page params>",
  "value": "[{...data...}]",
}
```

How I can tell infinity use the value of `@odata.nextLink` for the next page unless is empty/not-exists?

Docs: https://learn.microsoft.com/en-us/graph/paging?tabs=http#how-paging-works

---

The Azure Cost API works on the same way (a page Link is embedded inside the response):

https://github.com/grafana/grafana-infinity-datasource/blob/8547209d10d9b41d13d5e682036692dde2177ff4/testdata/misc/azure-cost-management-daily.json#L9

**Comment by xaibohphobiax** [comment]
@jkroepke Did you find a solution for this?

**Reply by jkroepke** [reply]
Yes, we implement a proxy between infinity datasource and MS Graph API and the proxy does the pagination for us.

initially, we solved without an proxy, but infinity datasource have builting limit of 5 pages at max. Ref: https://github.com/grafana/grafana-infinity-datasource/discussions/601#discussioncomment-11382097

Maybe @cjedro can assist here.

**Reply by cjedro** [reply]
correct, we added a proxy that not only does pagination but also authentication with workload identity instead of client secret. Basically the proxy can add additional logic that is not feasible/ too much complexity with infinity data source

Let me know if you got additional questions. Currently our proxy is closed source but open sourcing it is on my to do list. Some code cleanup and general review is required

**Reply by xaibohphobiax** [reply]
Thanks for the quick response!
How did you solve pagination without the proxy initially? Proxy idea sounds interesting too.

**Reply by jkroepke** [reply]
Thats 2 years ago, I forgot it.

**Reply by yesoreyeram** [reply]
@cjedro - interesting. Can you share the workload identify auth setup? I am working on something similar to natively bringing that to infinity and other grafana plugins.

**Reply by jkroepke** [reply]
@yesoreyeram The Grafana Azure SDK (which is used by core and other plugins) already supports Workload Identitiy. 

Ref: https://github.com/grafana/grafana-azure-sdk-go/blob/main/aztokenprovider/retriever_workloadidentity.go

See also: https://github.com/grafana/grafana-infinity-datasource/pull/785
